### PR TITLE
fix [GreaterThan,LessThan,Equals] HavingSpecs 

### PR DESCRIPTION
**Pull request by himanshug**
to do the comparison based on floats and longs appropriately.

See https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/druid-user/n9x5YyhRcl4/SnvSDyqOCwAJ

### CODE REVIEW

File: processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  static int compare(Row row, String aggregationName, Number value)
+  {
+    Object metricValueObj = row.getRaw(aggregationName);
+    if (metricValueObj != null) {
+      if (metricValueObj instanceof Long) {
+        long l = ((Long) metricValueObj).longValue();
+        return Longs.compare(l, value.longValue());
+      } else if (metricValueObj instanceof String) {
+        String metricValueStr = (String) metricValueObj;
+        if (!metricValueStr.contains(".")) {
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java]: 1E10 is a valid float representation, right?

**Comment by drcrallen** [comment]
Is there any performance degradation due to the extra checks?

File: processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java (hunk 1 above)
Comment (himanshug) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java]: true, fixed.

**Comment by himanshug** [comment]
@drcrallen  haven't done any perf testing but it appears that checks are hard to avoid in this situation to fix the bug.

**Comment by drcrallen** [comment]
@himanshug I looked at the way groupBy queries work and I think you are correct. Since they are row-based instead of column-based the filtering methods used elsewhere don't immediately apply.

File: processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        long l = ((Long) metricValueObj).longValue();
+        return Longs.compare(l, value.longValue());
+      } else if (metricValueObj instanceof String) {
+        String metricValueStr = (String) metricValueObj;
+        if (LONG_PAT.matcher(metricValueStr).matches()) {
+          long l = row.getLongMetric(aggregationName);
+          return Longs.compare(l, value.longValue());
+        }
+      }
+    }
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java]: can we throw an exception if the type is something else unknown?

**Comment by fjy** [comment]
:+1:

File: processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java (hunk 2 above)
Comment (himanshug) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java]: that would be breakward incompatible. So, I am defaulting to floats in the unknown case.

File: processing/src/test/java/io/druid/query/groupby/having/HavingSpecTest.java [hunk 3]
```diff
@@ -88,14 +88,17 @@ public void testHavingClauseSerde() throws Exception {
...
+    assertFalse(spec.eval(getTestRow(100.05f)));
+
+    spec = new GreaterThanHavingSpec("metric", 100.56f);
+    assertFalse(spec.eval(getTestRow(100.56f)));
+    assertFalse(spec.eval(getTestRow(90.53f)));
+    assertTrue(spec.eval(getTestRow(101.34f)));
+    assertTrue(spec.eval(getTestRow(Long.MAX_VALUE)));
   }
 
   private MapBasedInputRow makeRow(long ts, String dim, int value)
```
Comment (guobingkun) [inline processing/src/test/java/io/druid/query/groupby/having/HavingSpecTest.java]: This doesn't seem to be used anywhere.

Comment (himanshug) [inline processing/src/test/java/io/druid/query/groupby/having/HavingSpecTest.java]: removed.

File: processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import com.google.common.primitives.Longs;
+import io.druid.data.input.Row;
+
+import java.util.regex.Pattern;
+
+/**
+ */
+class HavingSpecMetricComparator
+{
+  private static final Pattern LONG_PAT = Pattern.compile("[-|+]?\\d+");
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java]: Can you make this package protected so you can test it explicitly?

specifically I would like to see a test for matching against `3456.3` or similar

Comment (drcrallen) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java]: And other general use cases :)

Comment (himanshug) [inline processing/src/main/java/io/druid/query/groupby/having/HavingSpecMetricComparator.java]: added

**Comment by drcrallen** [comment]
@himanshug can you cherry-pick a backport PR into the 0.8.2 branch?

**Comment by himanshug** [comment]
@drcrallen done

**Pull request merged**